### PR TITLE
Implemented the ContentButton

### LINKED ISSUE #383: [Enhancement] A button with content layout
## Summary

Xamarin forms Button control is limited to text and icon. Usually we used to create components with Frame or Grid if we require a to define a Button with complex template.
## API Changes

e.g.

```
<views:ContentButton Padding="2" BorderColor="Red" BackgroundColor="Gray" CornerRadius="12" IsClippedToBounds="True"  WidthRequest="200">
                <views:ContentButton.Content>
                        <StackLayout Orientation="Vertical" HorizontalOptions="FillAndExpand" VerticalOptions="FillAndExpand">
                            <Label Text="Test Button" TextColor="Black" HorizontalOptions="Center" VerticalOptions="Center" FontSize="Medium" />
                             <Label Text="Test subheading" TextColor="Black" HorizontalOptions="Center" VerticalOptions="Center" FontSize="Small" />
                        </StackLayout>
                </views:ContentButton.Content>
            </views:ContentButton>
```

## Who Will Do The Work?

- [x] I am willing to take this on myself

### PR DESCRIPTION
**Pull request by levelfivecoder**
### Description of Change ###

Xamarin forms Button control is limited to text and icon. Usually we used to create components with Frame or Grid if we require a to define a Button with complex template.

## Features ##

-Fixes #383

- A new button control that is capable of adding UI templates
- CornerRadius & BorderColor
- Clicked event & Command

##API Changes##

Using the control
Create an instance for the ContentButton control

```
<views:ContentButton Margin="{StaticResource ContentPadding}"
                CornerRadius="8"
                Clicked="ContentButton_Clicked"
                BackgroundColor="White"
                BorderColor="#8a8a8a">
            <views:ContentButton.Content>
                <Grid ColumnSpacing="20">
                    <Grid.ColumnDefinitions>
                        <ColumnDefinition Width="12" />
                        <ColumnDefinition Width="*" />
                    </Grid.ColumnDefinitions>
                    <BoxView BackgroundColor="#00ab64" />
                    <StackLayout Grid.Column="1" Padding="0,24,24,24"
                            Spacing="8">
                        <Label Style="{StaticResource label_section_header}"
                                Text="Button Title" />
                        <Label Text="Button description" />
                    </StackLayout>
                </Grid>
            </views:ContentButton.Content>
        </views:ContentButton>
```

### PR Checklist ###

- [ ] Has tests (if omitted, state reason in description)
- [✔]  Has samples (if omitted, state reason in description)
- [ ] Rebased on top of main at time of PR
- [✔]  Changes adhere to coding standard

### CODE REVIEW

File: XamarinCommunityToolkitSample/Resx/AppResources.es.resx [hunk 1]
```diff
@@ -473,4 +473,16 @@ If the user manually assigns a row or column on a view, it will be honored.</val
   <data name="XAMLExtension" xml:space="preserve">
     <value>#XAML Extension</value>
   </data>
+  <data name="ContentButtonTitle" xml:space="preserve">
+    <value>ContentButton</value>
+  </data>
+  <data name="ContentButtonDescription" xml:space="preserve">
+    <value>ContentButton alows you to add content view to a button</value>
```
Comment (VladislavAntonyuk) [inline XamarinCommunityToolkitSample/Resx/AppResources.es.resx:480]: typo allows

File: XamarinCommunityToolkitSample/Resx/AppResources.resx [hunk 2]
```diff
@@ -497,4 +497,16 @@ If the user manually assigns a row or column on a view, it will be honored.</val
   <data name="XAMLExtension" xml:space="preserve">
     <value>XAML Extension</value>
   </data>
+  <data name="ContentButtonTitle" xml:space="preserve">
+    <value>ContentButton</value>
+  </data>
+  <data name="ContentButtonDescription" xml:space="preserve">
+    <value>ContentButton alows you to add content view to a button</value>
```
Comment (VladislavAntonyuk) [inline XamarinCommunityToolkitSample/Resx/AppResources.resx:504]: typo allows

**Review by AndreiMisiukevich — CHANGES_REQUESTED** [review]
Hm @Char0394 could you please take a look at this PR
We have something similar https://github.com/xamarin/XamarinCommunityToolkit/blob/main/XamarinCommunityToolkit/Behaviors/Animations/AnimationBehavior.shared.cs

File: XamarinCommunityToolkit/Views/ContentButton.shared.cs [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+			set => SetValue(CommandParameterProperty, value);
+		}
+
+		public ICommand Command
+		{
+			get => (ICommand)GetValue(CommandProperty);
+			set => SetValue(CommandProperty, value);
+		}
+
+		ContentView contentHolder;
```
[the remark this answers is not in the document]
Comment (levelfivecoder) [reply XamarinCommunityToolkit/Views/ContentButton.shared.cs]: removed contentHolder

File: XamarinCommunityToolkitSample/Xamarin.CommunityToolkit.Sample.csproj [hunk 4]
```diff
@@ -12,18 +12,10 @@
...
     </Compile>
-    <Compile Update="Resx\AppResources.es.Designer.cs">
-      <DesignTime>True</DesignTime>
-      <AutoGen>True</AutoGen>
-      <DependentUpon>AppResources.es.resx</DependentUpon>
-    </Compile>
+    <EmbeddedResource Remove="Resx\AppResources.es.resx" />
   </ItemGroup>
 
   <ItemGroup>
```
[the remark this answers is not in the document]
Comment (levelfivecoder) [reply XamarinCommunityToolkitSample/Xamarin.CommunityToolkit.Sample.csproj]: revert back to the old file

File: XamarinCommunityToolkitSample/Pages/Views/ContentButtonPage.xaml.cs [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+using Xamarin.CommunityToolkit.Sample.Resx;
+
+namespace Xamarin.CommunityToolkit.Sample.Pages.Views
+{
+	public partial class ContentButtonPage : BasePage
+	{
+		public ContentButtonPage()
+			=> InitializeComponent();
+
+		async void ContentButton_Clicked(System.Object sender, System.EventArgs e)
```
[the remark this answers is not in the document]
Comment (levelfivecoder) [reply XamarinCommunityToolkitSample/Pages/Views/ContentButtonPage.xaml.cs]: Removed button click in page code behind and implemented command binding with viewmodel (included the counter)

File: XamarinCommunityToolkitSample/Pages/Views/ContentButtonPage.xaml.cs [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+		{
+			var result = await this.DisplaySnackbar(AppResources.ContentButtonTapMessage, 3000, "Run action", () =>
```
[the remark this answers is not in the document]
Comment (levelfivecoder) [reply XamarinCommunityToolkitSample/Pages/Views/ContentButtonPage.xaml.cs]: implemented Clicks counter

File: XamarinCommunityToolkitSample/ViewModels/Views/ContentButtonViewModel.cs [hunk 7]
```diff
@@ -0,0 +1,9 @@
+﻿using System.Windows.Input;
+using Xamarin.Forms;
+
+namespace Xamarin.CommunityToolkit.Sample.ViewModels.Views
+{
+	public class ContentButtonViewModel : BaseViewModel
+	{
+	}
```
[the remark this answers is not in the document]
Comment (levelfivecoder) [reply XamarinCommunityToolkitSample/ViewModels/Views/ContentButtonViewModel.cs:22]: Added command and counter logics

**Comment by AndreiMisiukevich** [comment]
Is there any advantage of this control over using AnimationBehavior with a Frame? 

https://github.com/xamarin/XamarinCommunityToolkit/blob/main/XamarinCommunityToolkit/Behaviors/Animations/AnimationBehavior.shared.cs

**Comment by domagojmedo** [comment]
@AndreiMisiukevich to me this is much nicer syntax to use than AnimationBehavior. Also it doesn't seem intuitive to me that you should use AnimationBehaviour if you want to accomplish this

**Comment by levelfivecoder** [comment]
> Is there any advantage of this control over using AnimationBehavior with a Frame?
> 
> https://github.com/xamarin/XamarinCommunityToolkit/blob/main/XamarinCommunityToolkit/Behaviors/Animations/AnimationBehavior.shared.cs

this provide a better syntax, instead of behaviour we can use as a VIew. Also we must think of animating the contentbutton by adding a bindable property with different animation options (This I require help)

**Comment by levelfivecoder** [comment]
Any update on this?

File: XamarinCommunityToolkit/Views/ContentButton.shared.cs [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+			remove => clickedEventManager.RemoveEventHandler(value);
+		}
+
+		public static readonly BindableProperty ContentTemplateProperty
+			= BindableProperty.Create(nameof(ContentTemplate), typeof(DataTemplate), typeof(ContentButton), propertyChanged: OnContentTemplatePropertyChanged);
+
+		public static readonly BindableProperty ContentProperty
+			= BindableProperty.Create(nameof(Content), typeof(View), typeof(ContentButton), propertyChanged: OnContentPropertyChanged);
+
+		public static readonly BindableProperty CornerRadiusProperty = BindableProperty.Create(nameof(CornerRadius), typeof(float), typeof(ContentButton), 0.0f, propertyChanged: OnCornerRadiusPropertyChanged);
```
Comment (AndreiMisiukevich) [inline XamarinCommunityToolkit/Views/ContentButton.shared.cs:24]:
```suggestion
		public static readonly BindableProperty CornerRadiusProperty 
			= BindableProperty.Create(nameof(CornerRadius), typeof(float), typeof(ContentButton), 0.0f, propertyChanged: OnCornerRadiusPropertyChanged);
```

File: XamarinCommunityToolkit/Views/ContentButton.shared.cs [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+
+		public static readonly BindableProperty BorderColorProperty = BindableProperty.Create(nameof(BorderColor), typeof(Color), typeof(ContentButton), Color.Default, propertyChanged: OnBorderColorPropertyChanged);
```
Comment (AndreiMisiukevich) [inline XamarinCommunityToolkit/Views/ContentButton.shared.cs:26]:
```suggestion
		public static readonly BindableProperty BorderColorProperty 
			= BindableProperty.Create(nameof(BorderColor), typeof(Color), typeof(ContentButton), Color.Default, propertyChanged: OnBorderColorPropertyChanged);
```

File: XamarinCommunityToolkit/Views/ContentButton.shared.cs [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+
+		public static new readonly BindableProperty BackgroundColorProperty = BindableProperty.Create(nameof(BackgroundColor), typeof(Color), typeof(ContentButton), Color.Default, propertyChanged: OnBackgroundColorPropertyChanged);
```
Comment (AndreiMisiukevich) [inline XamarinCommunityToolkit/Views/ContentButton.shared.cs:28]:
```suggestion
		public static new readonly BindableProperty BackgroundColorProperty = BindableProperty.Create(nameof(BackgroundColor), typeof(Color), typeof(ContentButton), Color.Default, propertyChanged: OnBackgroundColorPropertyChanged);
```
```suggestion
		public static new readonly BindableProperty BackgroundColorProperty 
			= BindableProperty.Create(nameof(BackgroundColor), typeof(Color), typeof(ContentButton), Color.Default, propertyChanged: OnBackgroundColorPropertyChanged);
```

File: XamarinCommunityToolkit/Views/ContentButton.shared.cs [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+
+		void OnBorderColorPropertyChanged()
+		{
+			Control.BorderColor = BorderColor;
+		}
+
+		void OnBackgroundColorPropertyChanged()
+		{
+			Control.BackgroundColor = BackgroundColor;
+		}
```
Comment (AndreiMisiukevich) [inline XamarinCommunityToolkit/Views/ContentButton.shared.cs:70]:
```suggestion
			=> Control.CornerRadius = CornerRadius;

		void OnBorderColorPropertyChanged()
			=> Control.BorderColor = BorderColor;

		void OnBackgroundColorPropertyChanged()
			=> Control.BackgroundColor = BackgroundColor;
```

File: XamarinCommunityToolkit/Views/ContentButton.shared.cs [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+			if (template == previousTemplate && Content != null)
+				return null;
+
+			previousTemplate = template;
+			return (View)template?.CreateContent();
+		}
+
+		protected override void OnControlInitialized(Frame control)
+		{
+			Control.GestureRecognizers.Add(new TapGestureRecognizer
```
Comment (AndreiMisiukevich) [inline XamarinCommunityToolkit/Views/ContentButton.shared.cs:163]:
```suggestion
			control.Padding = 0;
			control.GestureRecognizers.Add(new TapGestureRecognizer
```

File: XamarinCommunityToolkit/Views/ContentButton.shared.cs [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+			}
+		}
+
+		void SetContent()
+		{
+			if (Content != null)
+			{
+				Control.Padding = new Thickness(0);
+				Control.Content = Content;
+			}
```
Comment (AndreiMisiukevich) [inline XamarinCommunityToolkit/Views/ContentButton.shared.cs:143]:
```suggestion
				Control.Content = Content;
```

**Review by AndreiMisiukevich — COMMENTED** [review]
Sorry for the delay

File: XamarinCommunityToolkit/Views/ContentButton.shared.cs [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+
+					shouldIgnoreContentSetting = false;
+				}
+			}
+		}
+
+		void SetContent()
+		{
+			if (Content != null)
+				Control.Content = Content;
```
Comment (AndreiMisiukevich) [inline XamarinCommunityToolkit/Views/ContentButton.shared.cs:137]:
```suggestion
				Control.Content = Content;
				
```

File: XamarinCommunityToolkit/Views/ContentButton.shared.cs [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+
+		protected override void OnControlInitialized(Frame control)
+		{
+			control.Padding = 0;
+			control.GestureRecognizers.Add(new TapGestureRecognizer
+			{
+				Command = new Command(async (parameter) =>
+				{
+					await Control.FadeTo(0.75, 100);
+					await Control.FadeTo(1, 100);
```
Comment (AndreiMisiukevich) [inline XamarinCommunityToolkit/Views/ContentButton.shared.cs:164]: should be converted to single animation with two nested ones.

Child Animation: https://www.thewissen.io/playing-with-animation-in-xamarin-forms/
(the first animation will be from 0 to .5, the second one from .5 till 1. Animation time will be 200)

**Review by AndreiMisiukevich — DISMISSED** [review]
Please fix merge conflicts

**Comment by levelfivecoder** [comment]
> Please fix merge conflicts

conflicts resolved

File: XamarinCommunityToolkit/Views/ContentButton.shared.cs (hunk 15 above)
Comment (AndreiMisiukevich) [reply XamarinCommunityToolkit/Views/ContentButton.shared.cs:164]: I am sure it should be converted to a single animation

**Comment by jsuarezruiz** [comment]
Sorry @levelfivecoder there are more conflicts. Could you fix the merge conflicts?. Let me know if need help with anything.

**Comment by levelfivecoder** [comment]
> Sorry @levelfivecoder there are more conflicts. Could you fix the merge conflicts?. Let me know if need help with anything.

@jsuarezruiz Sorry I was busy with my office works and travel for the past two weeks, So I was not able to check this PR. I've resolved the merge conflicts. Please check

File: XamarinCommunityToolkit/Views/ContentButton.shared.cs [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+			remove => clickedEventManager.RemoveEventHandler(value);
+		}
+
+		public static readonly BindableProperty ContentTemplateProperty
+			= BindableProperty.Create(nameof(ContentTemplate), typeof(DataTemplate), typeof(ContentButton), propertyChanged: OnContentTemplatePropertyChanged);
+
+		public static readonly BindableProperty ContentProperty
+			= BindableProperty.Create(nameof(Content), typeof(View), typeof(ContentButton), propertyChanged: OnContentPropertyChanged);
+
+		public static readonly BindableProperty CornerRadiusProperty
```
[the remark this answers is not in the document]
Comment (YZahringer) [reply XamarinCommunityToolkit/Views/ContentButton.shared.cs:24]: [Frame.CornerRadius](https://docs.microsoft.com/en-us/dotnet/api/xamarin.forms.frame.cornerradius?view=xamarin-forms) does not allow it. Maybe clip it with [RoundRectangleGeometry](https://github.com/xamarin/Xamarin.Forms/pull/11851)?

[the remark this answers is not in the document]
Comment (AndreiMisiukevich) [reply XamarinCommunityToolkit/Views/ContentButton.shared.cs:24]: It was introduced in XF 5.0
I guess, we support lower versions as well

File: XamarinCommunityToolkit/Views/ContentButton.shared.cs [hunk 17]
```diff
@@ -0,0 +1,173 @@
...
+		}
+
+		protected override void OnControlInitialized(Frame control)
+		{
+			control.Padding = 0;
+			control.GestureRecognizers.Add(new TapGestureRecognizer
+			{
+				Command = new Command(async (parameter) =>
+				{
+					await Control.FadeTo(0.75, 100);
```
[the remark this answers is not in the document]
Comment (levelfivecoder) [reply XamarinCommunityToolkit/Views/ContentButton.shared.cs:163]: Can we add animations in the next version, may be I need some help from others also.

[the remark this answers is not in the document]
Comment (levelfivecoder) [reply XamarinCommunityToolkit/Views/ContentButton.shared.cs:163]: > There is only `Pressed` animation defined, is there a plan to support `Hover` animation? Maybe based on [TouchEffect](https://github.com/AndreiMisiukevich/TouchEffect)?

Can we add animations in next versions?, may be i need some help from others. I've plans to create bindable property for  selection of animations using enum.

[the remark this answers is not in the document]
Comment (levelfivecoder) [reply XamarinCommunityToolkit/Views/ContentButton.shared.cs:163]: not right now, we can include in the future stages

File: XamarinCommunityToolkit/Views/ContentButton.shared.cs (hunk 16 above)
[the remark this answers is not in the document]
Comment (levelfivecoder) [reply XamarinCommunityToolkit/Views/ContentButton.shared.cs:24]: @AndreiMisiukevich Should we change?

[the remark this answers is not in the document]
Comment (VladislavAntonyuk) [reply XamarinCommunityToolkit/Views/ContentButton.shared.cs:24]: I saw a PR to update to XF 5.0, and looks like it is a plan for the future. Waiting for the stable release.

**Comment by jsuarezruiz** [comment]
@levelfivecoder Could you rebase to fix the conflicts?

**Comment by Cfun1** [comment]
Is this pr still alive ?

**Comment by pictos** [comment]
@levelfivecoder are you able to move this PR forward?

**Comment by levelfivecoder** [comment]
> @levelfivecoder are you able to move this PR forward?

Sorry I was busy with my works in recent months, I had tried to rebase this branch. due to some conflicts I couldn't do that, Can someone rebase this?

**Comment by pictos** [comment]
> > @levelfivecoder are you able to move this PR forward?
> 
> Sorry I was busy with my works in recent months, I had tried to rebase this branch. due to some conflicts I couldn't do that, Can someone rebase this?

Sure thing! But after sync will you be able to move this forward or should we do it?

**Comment by levelfivecoder** [comment]
> > > @levelfivecoder are you able to move this PR forward?
> > 
> > 
> > Sorry I was busy with my works in recent months, I had tried to rebase this branch. due to some conflicts I couldn't do that, Can someone rebase this?
> 
> Sure thing! But after sync will you be able to move this forward or should we do it?

Thanks, hope I'll get some time. I really wanted to move forward with this PR

**Comment by pictos** [comment]
> Thanks, hope I'll get some time. I really wanted to move forward with this PR

@levelfivecoder if you need anything, please let me know

**Comment by levelfivecoder** [comment]
> > Thanks, hope I'll get some time. I really wanted to move forward with this PR
> 
> @levelfivecoder if you need anything, please let me know

HI Can you please follow up the PR, I'm too busy with my office works these days

**Comment by levelfivecoder** [comment]
@pictos Can you please follow up the PR, I'm too busy with my office works these days

**Comment by pictos** [comment]
@levelfivecoder sure thing, thanks for your work (:

**Comment by pictos** [comment]
@levelfivecoder when you have time, can you give me context on what is missing?

File: XamarinCommunityToolkit/Views/ContentButton.shared.cs (hunk 17 above)
[the remark this answers is not in the document]
Comment (pictos) [reply XamarinCommunityToolkit/Views/ContentButton.shared.cs]: @AndreiMisiukevich do you think that is a good idea to use the TouchEffect to handle the click action? Maybe I can refactor it and expose some props to devs.

**Comment by brminnick** [comment]
Closing PR as it has been abandoned.

If anybody is interested in taking over this PR we are more than happy to re-open it!

**Pull request closed without merging**